File: Java/Problem4-ShortestPalindrome.java
import java.math.BigInteger;

class Solution {
    public String shortestPalindrome(String s) {
        BigInteger base = new BigInteger("31");
        BigInteger mod = new BigInteger("1000000007"); //1e9 + 7

        BigInteger hashF = new BigInteger("0");
        BigInteger hashR = new BigInteger("0");
        BigInteger power = new BigInteger("1");
        int maxLength = 0;

        int i = 0;
        for (char c : s.toCharArray()){
            // int val = c - 'a' + 1;
            // BigInterger ch = new BigInterger(String.valueOf(val));
            BigInteger ch = BigInteger.valueOf(c - 'a' + 1);

            hashF = hashF.multiply(base).add(ch).mod(mod);
            hashR = hashR.add(power.multiply(ch)).mod(mod);

            power = power.multiply(base).mod(mod);

            if(hashF.equals(hashR)){
                maxLength = i + 1;
            }
            i++;
        }

        String suffix = s.substring(maxLength);
        StringBuilder sb = new StringBuilder(suffix);
        sb.reverse();
        suffix = sb.toString();

        return suffix + s;
    }
}

File: Python/Problem4-ShortestPalindrome.py
class Solution(object):
    def shortestPalindrome(self, s):
        base = 31
        mod = int(1e9) + 7
        
        hashF = 0
        hashR = 0
        power = 1
        maxLength = 0

        i = 0;
        for ch in s:
            ch = ord(ch) - ord('a') + 1

            hashF = (hashF * base + ch) % mod
            hashR = (hashR + power * ch) % mod

            power = (power * base) % mod

            if hashF == hashR:
                maxLength = i + 1

            i += 1

        suffix = s[maxLength:]
        suffix = suffix[::-1]
        
        return suffix + s
